File: MetaGeta/MetaGeta.DataStore/DataStore/DataStore.vb
<Serializable()> _
Public Class MGDataStore
    Implements INotifyPropertyChanged

    Private Shared ReadOnly log As log4net.ILog = log4net.LogManager.GetLogger(Reflection.MethodBase.GetCurrentMethod().DeclaringType)

    Private ReadOnly m_Owner As IDataStoreOwner
    Private ReadOnly m_DataMapper As IDataMapper
    Private m_ID As Long = -1
    Private m_Name As String
    Private m_Template As IDataStoreTemplate
    Private m_Description As String = ""

    Private ReadOnly m_AllPlugins As New List(Of IMGPluginBase)
    Private ReadOnly m_FileActionDictionary As New Dictionary(Of String, IMGFileActionPlugin)

    Private ReadOnly m_ShutdownEvent As New ManualResetEvent(False)

    Friend Sub New(ByVal owner As IDataStoreOwner, ByVal dataMapper As IDataMapper)
        m_Owner = owner
        m_DataMapper = dataMapper

        m_ImportThread.IsBackground = True
        m_ImportThread.Start()
    End Sub

    Public Sub Close()
        m_ShutdownEvent.Set()
        m_ImportThread.Join()
        For Each plugin In m_AllPlugins
            plugin.Shutdown()
        Next
    End Sub

    Public Sub Delete()
        m_Owner.DeleteDataStore(Me)
    End Sub

    Public Function GetAllTagOnFiles(ByVal tagName As String) As IList(Of Tuple(Of MGTag, MGFile))
        Return m_DataMapper.GetTagOnAllFiles(Me, tagName)
    End Function

    Public ReadOnly Property Files() As IList(Of MGFile)
        Get
            Return m_DataMapper.GetFiles(Me)
        End Get
    End Property

    Public Sub RemoveFiles(ByVal files As IEnumerable(Of MGFile))
        m_DataMapper.RemoveFiles(files, Me)
        OnFilesChanged()
    End Sub

#Region "Plugins"

    Friend Sub AddExistingPlugin(ByVal plugin As IMGPluginBase, ByVal id As Long)
        m_AllPlugins.Add(plugin)
        StartupPlugin(plugin, id, False)
    End Sub

    Friend Sub AddNewPlugin(ByVal pluginTypeName As String)
        Dim t = Type.GetType(pluginTypeName)
        If t Is Nothing Then
            Throw New Exception(String.Format("Can't find type ""{0}"".", pluginTypeName))
        End If
        Dim plugin = CType(Activator.CreateInstance(t), IMGPluginBase)
        m_AllPlugins.Add(plugin)
        'Datamapper will call StartupPlugin() after IDs have been assigned
    End Sub

    Friend Sub StartupPlugin(ByVal plugin As IMGPluginBase, ByVal id As Long, ByVal firstRun As Boolean)
        If Not Plugins.Contains(plugin) Then
            Throw New Exception("Can't find plugin in StartupPlugin().")
        End If

        log.InfoFormat("StartupPlugin(): {0}", plugin.GetType().FullName)

        plugin.Startup(Me, id)

        Dim settings = SettingInfoCollection.GetSettings(plugin)
        For Each setting In settings
            If firstRun Then
                setting.Value = setting.Metadata.DefaultValue
                SetPluginSetting(plugin, setting.SettingName, setting.GetDefaultValueAsString())
            Else
                setting.SetValueAsString(GetPluginSetting(plugin, setting.SettingName))
            End If
        Next
        AddHandler plugin.SettingChanged, AddressOf Plugin_SettingChanged

        If TypeOf plugin Is IMGFileActionPlugin Then
            SetUpAction(CType(plugin, IMGFileActionPlugin))
        End If
    End Sub

    Private Sub SetUpAction(ByVal ap As IMGFileActionPlugin)
        For Each s In ap.GetActions()
            m_FileActionDictionary.Add(s, ap)
        Next
    End Sub

    Public ReadOnly Property TaggingPlugins() As IEnumerable(Of IMGTaggingPlugin)
        Get
            Return m_AllPlugins.Where(Function(p) TypeOf p Is IMGTaggingPlugin).Cast(Of IMGTaggingPlugin)()
        End Get
    End Property
    Public ReadOnly Property FileSourcePlugins() As IEnumerable(Of IMGFileSourcePlugin)
        Get
            Return m_AllPlugins.Where(Function(p) TypeOf p Is IMGFileSourcePlugin).Cast(Of IMGFileSourcePlugin)()
        End Get
    End Property
    Public ReadOnly Property FileActionPlugins() As IEnumerable(Of IMGFileActionPlugin)
        Get
            Return m_AllPlugins.Where(Function(p) TypeOf p Is IMGFileActionPlugin).Cast(Of IMGFileActionPlugin)()
        End Get
    End Property

    Public ReadOnly Property Plugins() As IEnumerable(Of IMGPluginBase)
        Get
            Return m_AllPlugins
        End Get
    End Property

#End Region

#Region "Importing"

    Private ReadOnly m_ImportThread As New Thread(AddressOf ImportThread)
    Private ReadOnly m_ImportNow As New AutoResetEvent(False)

    Private m_LastImportStatus As New ImportStatus()
    Private m_LastImportStatusLock As New Object()

    ''' <summary>
    ''' Starts an asynchronous refresh and import from all FileSourcePlugins.
    ''' </summary>
    ''' <remarks></remarks>
    Public Sub BeginRefresh()
        m_ImportNow.Set()
    End Sub

    Public ReadOnly Property ImportStatus() As ImportStatus
        Get
            SyncLock m_LastImportStatusLock
                Return m_LastImportStatus
            End SyncLock
        End Get
    End Property

    Private Sub SetImportStatus(ByVal message As String, Optional ByVal progressPct As Integer = -1)
        SyncLock m_LastImportStatusLock
            If progressPct = -1 Then
                m_LastImportStatus = New ImportStatus(message)
            Else
                m_LastImportStatus = New ImportStatus(message, progressPct)
            End If
        End SyncLock
        OnPropertyChanged("ImportStatus")
    End Sub

    Public Sub ImportThread()
        While True
            Dim wakeupHandle = WaitHandle.WaitAny(New WaitHandle() {m_ImportNow, m_ShutdownEvent})
            If wakeupHandle = 1 Then Return

            SetImportStatus("Listing files...")

            'Get list of files
            Dim newPaths = (From fs In Me.FileSourcePlugins _
                      From path In fs.GetFilesToAdd() _
                      Select path).ToArray()
            Dim files = (From p In newPaths Select New MGFile(Me)).ToArray()
            m_DataMapper.WriteNewFiles(files, Me)

            For Each fileAndPath In files.IndexInnerJoin(newPaths)
                Dim newfile = fileAndPath.First
                Dim filePath = fileAndPath.Second
                Dim index = fileAndPath.Third
                SetImportStatus(String.Format("Importing ""{0}""", filePath.LocalPath), CInt(index / files.Length * 100))
                ImportNewFile(newfile, filePath)
            Next

            SetImportStatus("Import complete.", 100)

            OnFilesChanged()
        End While
    End Sub

    Public Function AddNewFile(ByVal path As Uri) As MGFile
        Dim file As New MGFile(Me)
        m_DataMapper.WriteNewFiles(file.SingleToEnumerable(), Me)
        ImportNewFile(file, path)
        OnFilesChanged()
        Return file
    End Function

    Private Sub ImportNewFile(ByVal newfile As MGFile, ByVal filePath As Uri)
        newfile.SetTag(MGFile.FileNameKey, filePath.AbsoluteUri)
        newfile.SetTag("ImportComplete", False.ToString())

        log.DebugFormat("Importing file: {0}....", filePath)

        For Each plugin As IMGTaggingPlugin In Me.TaggingPlugins
            log.DebugFormat("Importing {0} with {1}....", newfile.FileName, plugin.GetType().FullName)
            Try
                plugin.Process(newfile, New ProgressStatus())
            Catch ex As Exception
                log.Warn(String.Format("Importing ""{0}"" with {1} failed.", filePath, plugin.GetType().FullName), ex)
            End Try
            m_DataMapper.WriteFile(newfile)
        Next
    End Sub

    Private m_ImportQueueCache As ReadOnlyCollection(Of MGFile)
    Private ReadOnly Property ImportQueue() As IEnumerable(Of MGFile)
        Get
            If m_ImportQueueCache Is Nothing OrElse m_ImportQueueCache.Count = 0 Then
                m_ImportQueueCache = New ReadOnlyCollection(Of MGFile)(Files.Where(Function(f) Not Boolean.Parse(f.GetTag("ImportComplete").Coalesce("False"))).ToArray())
            End If
            Return m_ImportQueueCache
        End Get
    End Property

#End Region

#Region "Settings"
    Public Sub Plugin_SettingChanged(ByVal sender As Object, ByVal e As PropertyChangedEventArgs)
        Dim plugin = CType(sender, IMGPluginBase)
        Dim value = SettingInfoCollection.GetSettings(plugin).GetSetting(e.PropertyName).GetValueAsString()
        SetPluginSetting(plugin, e.PropertyName, value)
    End Sub

    Public Function GetPluginSetting(ByVal plugin As IMGPluginBase, ByVal settingName As String) As String
        Return m_DataMapper.GetPluginSetting(Me, plugin.PluginID, settingName)
    End Function
    Public Sub SetPluginSetting(ByVal plugin As IMGPluginBase, ByVal settingName As String, ByVal settingValue As String)
        m_DataMapper.WritePluginSetting(Me, plugin.PluginID, settingName, settingValue)
    End Sub
#End Region

    Public Property Name() As String
        Get
            Return m_Name
        End Get
        Set(ByVal value As String)
            If value <> m_Name Then
                m_Name = value
                If Not AreUpdatesSuspended Then
                    m_DataMapper.WriteDataStore(Me)
                    OnNameChanged()
                End If
            End If
        End Set
    End Property
    Public Property Description() As String
        Get
            Return m_Description
        End Get
        Set(ByVal value As String)
            If value <> m_Description Then
                m_Description = value
                If Not AreUpdatesSuspended Then
                    m_DataMapper.WriteDataStore(Me)
                    OnDescriptionChanged()
                End If
            End If
        End Set
    End Property
    Public Property Template() As IDataStoreTemplate
        Get
            Return m_Template
        End Get
        Set(ByVal value As IDataStoreTemplate)
            If value IsNot m_Template Then
                m_Template = value
                If Not AreUpdatesSuspended Then
                    m_DataMapper.WriteDataStore(Me)
                    OnTemplateChanged()
                End If
            End If
        End Set
    End Property
    Public Property ID() As Long
        Get
            Return m_ID
        End Get
        Set(ByVal value As Long)
            If value <> m_ID Then
                m_ID = value
            End If
        End Set
    End Property

#Region "Suspending Updates"

    Private m_SuspendCount As Integer = 0

    Private ReadOnly Property AreUpdatesSuspended() As Boolean
        Get
            Return m_SuspendCount <> 0
        End Get
    End Property

    Public Function SuspendUpdates() As IDisposable
        m_SuspendCount += 1
        Return New SuspendUpdatesToken(Me)
    End Function

    Private Class SuspendUpdatesToken
        Implements IDisposable

        Private ReadOnly m_DataStore As MGDataStore

        Public Sub New(ByVal dataStore As MGDataStore)
            m_DataStore = dataStore
        End Sub

        Private disposedValue As Boolean = False        ' To detect redundant calls

        ' IDisposable
        Protected Overridable Sub Dispose(ByVal disposing As Boolean)
            If Not Me.disposedValue Then
                If disposing Then
                    m_DataStore.m_SuspendCount -= 1
                End If

                ' TODO: free your own state (unmanaged objects).
                ' TODO: set large fields to null.
            End If
            Me.disposedValue = True
        End Sub

#Region " IDisposable Support "
        ' This code added by Visual Basic to correctly implement the disposable pattern.
        Public Sub Dispose() Implements IDisposable.Dispose
            ' Do not change this code.  Put cleanup code in Dispose(ByVal disposing As Boolean) above.
            Dispose(True)
            GC.SuppressFinalize(Me)
        End Sub
#End Region

    End Class

#End Region

    Public Overrides Function ToString() As String
        Return "MGDataStore: " & Name
    End Function

    Public Function ToCsv() As String
        Dim sb As New StringBuilder()
        sb.AppendLine(Aggregate s In Template.GetDimensionNames() Order By s Into JoinToCsv(s))
        For Each f In Me.Files
            sb.AppendLine(Aggregate t In f.Tags Order By t.Name Into JoinToCsv(t.Value))
        Next
        Return sb.ToString()
    End Function

#Region "Action Plugins"
    Public Sub DoAction(ByVal file As MGFile, ByVal action As String)
        m_Owner.EnqueueAction(action, Me, file)
    End Sub

    Friend Function LookupAction(ByVal action As String) As IMGFileActionPlugin
        Return m_FileActionDictionary(action)
    End Function
#End Region

#Region "Property Changed"

    Private Sub OnNameChanged()
        OnPropertyChanged("Name")
    End Sub
    Private Sub OnDescriptionChanged()
        OnPropertyChanged("Description")
    End Sub
    Private Sub OnTemplateChanged()
        OnPropertyChanged("Template")
    End Sub
    Private Sub OnFilesChanged()
        OnPropertyChanged("Files")
    End Sub

    Private Sub OnPropertyChanged(ByVal propertyName As String)
        RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs(propertyName))
    End Sub

    Public Event PropertyChanged(ByVal sender As Object, ByVal e As System.ComponentModel.PropertyChangedEventArgs) Implements System.ComponentModel.INotifyPropertyChanged.PropertyChanged
#End Region

#Region "DataStoreCreationArguments"

    Public Function GetCreationArguments() As DataStoreBuilder
        Return New DataStoreBuilder() With {.Name = Name, _
                                                      .Description = Description, _
                                                      .Tempate = Template, _
                                                      .DirectoriesToWatch = GetPluginSetting(CType(FileSourcePlugins.Single(), IMGPluginBase), "DirectoriesToWatch"), _
                                                      .Extensions = GetPluginSetting(CType(FileSourcePlugins.Single(), IMGPluginBase), "Extensions") _
                                                      }
    End Function

    Public Sub SetCreationArguments(ByVal args As DataStoreBuilder)
        Name = args.Name
        Description = args.Description
        SetPluginSetting(CType(FileSourcePlugins.Single(), IMGPluginBase), "DirectoriesToWatch", args.DirectoriesToWatch)
        SetPluginSetting(CType(FileSourcePlugins.Single(), IMGPluginBase), "Extensions", args.Extensions)
    End Sub

#End Region

#Region "Images"
    Public Function GetImageDirectory() As String
        Dim imagePath = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "Images")
        If Not Directory.Exists(imagePath) Then
            Directory.CreateDirectory(imagePath)
        End If
        Return imagePath
    End Function
#End Region
End Class

Public Class ImportStatus
    Private ReadOnly m_StatusMessage As String
    Private ReadOnly m_ProgressPct As Integer
    Private ReadOnly m_IsImporting As Boolean
    Private ReadOnly m_IsIndeterminate As Boolean

    Public Sub New(ByVal statusMessage As String, ByVal progressPct As Integer)
        m_StatusMessage = statusMessage
        m_ProgressPct = progressPct
        m_IsImporting = True
        m_IsIndeterminate = False
    End Sub

    Public Sub New()
        Me.New(String.Empty, 0)
        m_IsImporting = False
    End Sub

    Public Sub New(ByVal statusMessage As String)
        Me.New(statusMessage, 0)
        m_IsIndeterminate = True
    End Sub

    Public ReadOnly Property StatusMessage() As String
        Get
            Return m_StatusMessage
        End Get
    End Property

    Public ReadOnly Property ProgressPct() As Integer
        Get
            Return m_ProgressPct
        End Get
    End Property

    Public ReadOnly Property IsImporting() As Boolean
        Get
            Return m_IsImporting
        End Get
    End Property

    Public ReadOnly Property IsIndeterminate() As Boolean
        Get
            Return m_IsIndeterminate
        End Get
    End Property
End Class

File: MetaGeta/MetaGeta.DataStore/DataStore/DataStore.cs
// Copyright 2009 Will Thomas
// 
// This file is part of MetaGeta.
// 
// MetaGeta is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// MetaGeta is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with MetaGeta. If not, see <http://www.gnu.org/licenses/>.

#region

using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.ComponentModel;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Runtime.InteropServices;
using System.Text;
using System.Threading;
using log4net;
using MetaGeta.DataStore.Database;
using MetaGeta.Utilities;
using System.Concurrency;
using System.Threading.Tasks;

#endregion

namespace MetaGeta.DataStore {
	[Serializable()]
	public class MGDataStore : INotifyPropertyChanged {
		private static readonly ILog log = LogManager.GetLogger(MethodBase.GetCurrentMethod().DeclaringType);
		private readonly List<IMGPluginBase> m_AllPlugins = new List<IMGPluginBase>();
		private readonly IDataMapper m_DataMapper;
		private readonly IDataStoreOwner m_Owner;
		private readonly CancellationTokenSource m_CancelTaskSource = new CancellationTokenSource();
		private string m_Description = "";
		private long m_ID = -1;
		private string m_Name;
		private IDataStoreTemplate m_Template;

		internal MGDataStore(IDataStoreOwner owner, IDataMapper dataMapper) {
			m_Owner = owner;
			m_DataMapper = dataMapper;

			ImportStatus = new ImportStatus(string.Format("{0} files total.", m_DataMapper.GetFileCount(this)), 0);
		}

		public IList<MGFile> Files { get { return m_DataMapper.GetFiles(this); } }

		public string Name {
			get { return m_Name; }
			set {
				if (value != m_Name) {
					m_Name = value;
					if (!AreUpdatesSuspended) {
						m_DataMapper.WriteDataStore(this);
						OnNameChanged();
					}
				}
			}
		}

		public string Description {
			get { return m_Description; }
			set {
				if (value != m_Description) {
					m_Description = value;
					if (!AreUpdatesSuspended) {
						m_DataMapper.WriteDataStore(this);
						OnDescriptionChanged();
					}
				}
			}
		}

		public IDataStoreTemplate Template {
			get { return m_Template; }
			set {
				if (!ReferenceEquals(value, m_Template)) {
					m_Template = value;
					if (!AreUpdatesSuspended) {
						m_DataMapper.WriteDataStore(this);
						OnTemplateChanged();
					}
				}
			}
		}

		public long ID {
			get { return m_ID; }
			set {
				if (value != m_ID)
					m_ID = value;
			}
		}

		#region "Suspending Updates"

		private int m_SuspendCount = 0;

		private bool AreUpdatesSuspended { get { return m_SuspendCount != 0; } }

		public IDisposable SuspendUpdates() {
			m_SuspendCount += 1;
			return new SuspendUpdatesToken(this);
		}

		private class SuspendUpdatesToken : IDisposable {
			private readonly MGDataStore m_DataStore;

			// To detect redundant calls
			private bool disposedValue = false;

			public SuspendUpdatesToken(MGDataStore dataStore) {
				m_DataStore = dataStore;
			}

			// IDisposable
			protected virtual void Dispose(bool disposing) {
				if (!disposedValue) {
					if (disposing)
						m_DataStore.m_SuspendCount -= 1;

					// TODO: free your own state (unmanaged objects).
					// TODO: set large fields to null.
				}
				disposedValue = true;
			}

			#region " IDisposable Support "

			// This code added by Visual Basic to correctly implement the disposable pattern.
			public void Dispose() {
				// Do not change this code.  Put cleanup code in Dispose(ByVal disposing As Boolean) above.
				Dispose(true);
				GC.SuppressFinalize(this);
			}

			#endregion
		}

		#endregion

		public void Close() {
			m_CancelTaskSource.Cancel();
			foreach (IMGPluginBase plugin in m_AllPlugins)
				plugin.Shutdown();
		}

		public void Delete() {
			m_Owner.DeleteDataStore(this);
		}

		public IList<Tuple<MGTag, MGFile>> GetAllTagOnFiles(string tagName) {
			return m_DataMapper.GetTagOnAllFiles(this, tagName);
		}

		public void RemoveFiles(IEnumerable<MGFile> files) {
			m_DataMapper.RemoveFiles(files, this);
			OnFilesChanged();
		}

		public override string ToString() {
			return "MGDataStore: " + Name;
		}

		public string ToCsv() {
			var sb = new StringBuilder();
			sb.AppendLine();
			foreach (MGFile f in Files)
				sb.AppendLine();
			return sb.ToString();
		}

		#region "Action Plugins"

		public IEnumerable<IAction> GetActions() {
			return m_AllPlugins.OfType<IMGFileActionPlugin>().SelectMany(p => p.GetActions()).ToArray();
		}

		public void DoAction(MGFile file, IAction action) {
			m_Owner.EnqueueAction(action, this, file);
		}

		#endregion

		#region "Property Changed"

		public event PropertyChangedEventHandler PropertyChanged;

		private void OnNameChanged() {
			OnPropertyChanged("Name");
		}
		private void OnDescriptionChanged() {
			OnPropertyChanged("Description");
		}
		private void OnTemplateChanged() {
			OnPropertyChanged("Template");
		}
		private void OnFilesChanged() {
			OnPropertyChanged("Files");
		}
		private void OnPropertyChanged(string propertyName) {
			if (PropertyChanged != null)
				PropertyChanged(this, new PropertyChangedEventArgs(propertyName));
		}

		#endregion

		#region "DataStoreCreationArguments"

		public DataStoreBuilder GetCreationArguments() {
			var fileSourceSettings = SettingInfoCollection.GetSettings((IMGPluginBase)FileSourcePlugins.Single());
			return new DataStoreBuilder {
				Name = Name,
				Description = Description,
				Tempate = Template,
				DirectoriesToWatch = fileSourceSettings.GetSetting("DirectoriesToWatch").GetValueAsString(),
				Extensions = fileSourceSettings.GetSetting("Extensions").GetValueAsString()
			};
		}

		public void SetCreationArguments(DataStoreBuilder args) {
			Name = args.Name;
			Description = args.Description;
			var fileSourceSettings = SettingInfoCollection.GetSettings((IMGPluginBase)FileSourcePlugins.Single());
			fileSourceSettings.GetSetting("DirectoriesToWatch").SetValueAsString(args.DirectoriesToWatch);
			fileSourceSettings.GetSetting("Extensions").SetValueAsString(args.Extensions);
		}

		#endregion

		#region "Images"

		public string GetImageDirectory() {
			string imagePath = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "Images");
			if (!Directory.Exists(imagePath))
				Directory.CreateDirectory(imagePath);
			return imagePath;
		}

		#endregion

		#region "Plugins"

		public IEnumerable<IMGTaggingPlugin> TaggingPlugins {
			get { return m_AllPlugins.Where(p => p is IMGTaggingPlugin).Cast<IMGTaggingPlugin>(); }
		}

		public IEnumerable<IMGFileSourcePlugin> FileSourcePlugins {
			get { return m_AllPlugins.Where(p => p is IMGFileSourcePlugin).Cast<IMGFileSourcePlugin>(); }
		}

		public IEnumerable<IMGFileActionPlugin> FileActionPlugins {
			get { return m_AllPlugins.Where(p => p is IMGFileActionPlugin).Cast<IMGFileActionPlugin>(); }
		}

		public IEnumerable<IMGPluginBase> Plugins {
			get { return m_AllPlugins; }
		}

		internal void AddExistingPlugin(IMGPluginBase plugin, long id) {
			m_AllPlugins.Add(plugin);
			StartupPlugin(plugin, id, false);
		}

		internal void AddNewPlugin(string pluginTypeName) {
			Type t = Type.GetType(pluginTypeName);
			if (t == null)
				throw new Exception(string.Format("Can't find type \"{0}\".", pluginTypeName));
			var plugin = (IMGPluginBase)Activator.CreateInstance(t);
			m_AllPlugins.Add(plugin);
			//Datamapper will call StartupPlugin() after IDs have been assigned
		}

		internal void StartupPlugin(IMGPluginBase plugin, long id, bool firstRun) {
			if (!Plugins.Contains(plugin))
				throw new Exception("Can't find plugin in StartupPlugin().");

			log.InfoFormat("StartupPlugin(): {0}", plugin.GetType().FullName);

			if (plugin is IMGFileSourcePlugin) {
				var fsp = (IMGFileSourcePlugin)plugin;
				var bufferTime = TimeSpan.FromMilliseconds(100);
				fsp.FileAddedEvent.BufferWithTime(bufferTime).Where(l => l.Count > 0).Subscribe(l => OnFileAdded(fsp, l));
				fsp.FileModifiedEvent.BufferWithTime(bufferTime).Where(l => l.Count > 0).Subscribe(l => OnFileModified(fsp, l));
				fsp.FileMovedEvent.BufferWithTime(bufferTime).Where(l => l.Count > 0).Subscribe(l => OnFileMoved(fsp, l));
				fsp.FileDeleteEvent.BufferWithTime(bufferTime).Where(l => l.Count > 0).Subscribe(l => OnFileDeleted(fsp, l));
			}

			plugin.Startup(this, id);

			foreach (SettingInfo setting in SettingInfoCollection.GetSettings(plugin)) {
				if (firstRun) {
					setting.Value = setting.Metadata.DefaultValue;
					m_DataMapper.WritePluginSetting(this, plugin.PluginID, setting.SettingName, setting.GetDefaultValueAsString());
				} else
					setting.SetValueAsString(m_DataMapper.GetPluginSetting(this, plugin.PluginID, setting.SettingName));
			}
			plugin.SettingChanged += Plugin_SettingChanged;
		}

		#endregion

		#region "Importing"

		private ImportStatus m_LastImportStatus = ImportStatus.None;
		private object m_LastImportStatusLock = new object();

		public ImportStatus ImportStatus {
			get {
				lock (m_LastImportStatusLock) {
					return m_LastImportStatus;
				}
			}
			private set {
				lock (m_LastImportStatusLock) {
					m_LastImportStatus = value;
					OnPropertyChanged("ImportStatus");
				}
			}
		}


		/// <summary>
		/// Starts an asynchronous refresh and import from all FileSourcePlugins.
		/// </summary>
		/// <remarks></remarks>
		public void BeginRefresh() {
			ImportStatus = new ImportStatus("Listing files...", 0);
			FileSourcePlugins.AsParallel().WithCancellation(m_CancelTaskSource.Token).ForEach(fs => fs.Refresh(m_CancelTaskSource.Token));
		}

		private void OnFileAdded(IMGFileSourcePlugin pluing, IList<FileAddedEventArgs> events) {
			var files = events.Select(e => new MGFile(this)).ToArray();
			m_DataMapper.WriteNewFiles(files, this);
			var filePaths = events.Select(e => e.FilePath).ToArray();
			for (int i = 0; i < filePaths.Length; i++) {
				ImportStatus = new ImportStatus(string.Format("Importing \"{0}\"", filePaths[i].LocalPath), filePaths.Length - 1);
				ImportNewFile(files[i], filePaths[i]);
			}
			ImportStatus = new ImportStatus("Import complete.", 0);
			m_Owner.MainThreadScheduler.Schedule(OnFilesChanged);
		}
		private void OnFileModified(IMGFileSourcePlugin pluing, IList<FileModifiedEventArgs> events) {

		}
		private void OnFileMoved(IMGFileSourcePlugin pluing, IList<FileMovedEventArgs> events) {

		}
		private void OnFileDeleted(IMGFileSourcePlugin pluing, IList<FileDeletedEventArgs> events) {
			var removedFiles = events.ToLookup(e => e.FilePath);
			m_DataMapper.RemoveFiles(Files.Where(f => removedFiles.Contains(f.FileNameUri)), this);
			ImportStatus = new ImportStatus("", 0);
			m_Owner.MainThreadScheduler.Schedule(OnFilesChanged);
		}

		public MGFile AddNewFile(Uri path) {
			var file = new MGFile(this);
			m_DataMapper.WriteNewFiles(file.SingleToEnumerable(), this);
			ImportNewFile(file, path);
			m_Owner.MainThreadScheduler.Schedule(OnFilesChanged);
			return file;
		}

		private void ImportNewFile(MGFile newfile, Uri filePath) {
			newfile.Tags.Set(MGFile.FileNameKey, filePath.AbsoluteUri);

			log.DebugFormat("Importing file: {0}....", filePath);

			foreach (IMGTaggingPlugin plugin in TaggingPlugins) {
				log.DebugFormat("Importing {0} with {1}....", newfile.FileName, plugin.GetType().FullName);
				try {
					plugin.Process(newfile, new ProgressStatus());
				} catch (Exception ex) {
					log.Warn(string.Format("Importing \"{0}\" with {1} failed.", filePath, plugin.GetType().FullName), ex);
				}
				m_DataMapper.WriteFile(newfile);
			}
		}

		#endregion

		#region "Settings"

		public void Plugin_SettingChanged(object sender, PropertyChangedEventArgs e) {
			var plugin = (IMGPluginBase)sender;
			string value = SettingInfoCollection.GetSettings(plugin).GetSetting(e.PropertyName).GetValueAsString();
			m_DataMapper.WritePluginSetting(this, plugin.PluginID, e.PropertyName, value);
		}

		#endregion
	}

	public class ImportStatus {
		private readonly int m_FilesRemaining;
		private readonly string m_StatusMessage;

		public ImportStatus(string statusMessage, int filesRemaining) {
			m_StatusMessage = statusMessage;
			m_FilesRemaining = filesRemaining;
		}

		public string StatusMessage { get { return m_StatusMessage; } }
		public int FilesRemaining { get { return m_FilesRemaining; } }

		public static readonly ImportStatus None = new ImportStatus(string.Empty, 0);
	}
}